src/main/java/com/mongodb/spark/sql/connector/schema/SchemaToExpressionEncoderFunction.java

```diff
@@ -0,0 +1,112 @@
+import org.apache.spark.sql.catalyst.encoders.ExpressionEncoder;
+import org.apache.spark.sql.catalyst.expressions.Attribute;
+import org.apache.spark.sql.types.StructType;
+import scala.collection.immutable.Seq;
+
+@SuppressWarnings("unchecked")
+class SchemaToExpressionEncoderFunction
+    implements SerializableFunction<StructType, ExpressionEncoder<Row>> {
+  private static final long serialVersionUID = 1L;
+  private static final Object MODULE_STATIC;
+  private static final Method APPLY_METHOD;
+  private static final SerializableFunction<StructType, Seq<Attribute>> ATTRIBUTE_FUNCTION;
+
+  static {
+    Class<?> clazz;
+    Object moduleStatic = null;
+    Method applyMethod = null;
+    SerializableFunction<StructType, Seq<Attribute>> attributeFunction = null;
+
+    try {
+      clazz = Class.forName("org.apache.spark.sql.catalyst.encoders.RowEncoder$");
+      moduleStatic = clazz.getDeclaredField("MODULE$").get(null);
+      applyMethod = clazz.getMethod("apply", StructType.class);
+      Method toAttributesMethod = StructType.class.getMethod("toAttributes");
+      attributeFunction = schema -> MethodInvoker.invoke(toAttributesMethod, schema);
+    } catch (ClassNotFoundException
+        | NoSuchFieldException
+        | NoSuchMethodException
+        | IllegalAccessException e) {
+      // Ignore - not available in this version of Spark
+    }
+
+    if (applyMethod == null) {
+      try {
+        clazz = Class.forName("org.apache.spark.sql.catalyst.encoders.ExpressionEncoder$");
+        moduleStatic = clazz.getDeclaredField("MODULE$").get(null);
+        applyMethod = clazz.getMethod("apply", StructType.class);
+        Object dataTypeStatic = Class.forName("org.apache.spark.sql.catalyst.types.DataTypeUtils$")
+            .getDeclaredField("MODULE$")
+            .get(null);
+        Method toAttributesMethod =
+            dataTypeStatic.getClass().getMethod("toAttributes", StructType.class);
+        attributeFunction =
+            schema -> MethodInvoker.invoke(toAttributesMethod, dataTypeStatic, schema);
+      } catch (ClassNotFoundException
+          | NoSuchFieldException
+          | NoSuchMethodException
+          | IllegalAccessException e) {
+        // Ignore - not available in this version of Spark
+      }
+    }
+
+    if (applyMethod == null || attributeFunction == null) {
+      String sparkVersion = org.apache.spark.package$.MODULE$.SPARK_VERSION_SHORT();
+      throw new ExceptionInInitializerError("Unsupported version of Spark: " + sparkVersion);
+    }
+
+    MODULE_STATIC = moduleStatic;
+    APPLY_METHOD = applyMethod;
+    ATTRIBUTE_FUNCTION = attributeFunction;
+  }
+
+  @Override
+  public ExpressionEncoder<Row> apply(final StructType schema) {
+    ExpressionEncoder<Row> rowEncoder = MethodInvoker.invoke(APPLY_METHOD, MODULE_STATIC, schema);
+    return rowEncoder.resolveAndBind(ATTRIBUTE_FUNCTION.apply(schema), SimpleAnalyzer$.MODULE$);
```

Comment:
Get the resolved ExpressionEncoder which can create valid serializers and deserializers